### ISSUE TITLE
update rate management potential improvements

### ISSUE THREAD

**Issue by bltzr** [ossia/score#147]
as a follow-up of #146 - for future versions of i-score:

I don't think it's useful or desirable to have same update rates for all automations: sometimes some automations need to be very precise, sometimes they need to be very loose... most of the time, though (at least in my opinion) they don't need to be as quickly updating as they currently do (as of 1.0).

This update rate is depending on each automation purpose... so ideally, the update rate should be defined for each automation (also maybe for other processes ? I don't know..)
Though most of the the time, I think the update rate depends on the destination parameter (and all automations for this parameter might use the same update rate...)

So, my opinion to improve this would be:
- to have a default update rate for all automations (let's say 10 or 15 ms-
- to be able to change this rate for each automation (in the inspector)
- to be able to set a default, or preferred update rate in the device explorer tree for each parameter (so that, by default, every automation created for this parameter would happen at the chosen rate... and could also be changed when desired)
  Does that make sense ?

Of course, this is not a feature request I'm asking to ask upon urgently... I'm mostly raising this here in order for it to be discussed... and hopefully included in a future version of i-score...

On the implementation side (we've been discussing this with Théo this morning), it seems that the clock lays at the constraint level, so one option to handle this would be to keep the constraint clock at the highest possible level (i.e. 1ms for now... and maybe audio vectorsize if handling audio inside i-score ???), and subdivide the constraint clock for each automation, depending on the desired update rate...

I hope I didn't misunderstand things... please correct me (but not too hard !) if I did, Théo...

**Comment by bltzr** [comment]
opinions @jcelerier @nvuaille @reno- @jln- @cbossut @lossius ? (please cc to anyone you think might be interested...)

**Comment by reno-** [comment]
I totally agree that automation update rate need to be setable for each devices.
It will be really nice to have a different rate for each automation, but IMHO it could be done later on.

**Comment by bltzr** [comment]
yup, setting for a device could be a good start if that's simpler

**Comment by jln-** [comment]
@reno- Do you mean for the whole device or, as I understood from Pascal message, for each parameter of a device ? 

My use being mostly Max-related, I personally have little need for a device-based update rate. On the other hand, I do imagine the need for a parameter-based update rate.

Being able to do so on each automation certainly sounds nice but my feeling is that it's really really specific, isn't it ? Maybe even to a point that I feel its another use case (ie. writing some starcaise-like automation for example).

Out of curiosity, do you guys have encountered the need to have different update rates for some automations on a same parameter ? (I may certainly miss some practice here)

**Comment by bltzr** [comment]
> do you guys have encountered the need to have different update rates for some automations on a same parameter ?

I mainly encountered the problem of having my computer bloated by too much data coming in at 1KHz (current rate...)

**Comment by reno-** [comment]
You're right @jln- . Per-parameter sounds better than per-automation. But per-device is important too. For exemple if you have a serial device (with a osc2serial converter between) and an OSC device, you won't have the same rate. If you control DMX, 40 Hz could be enough and so on…

So maybe we should have per-device update rate and per-paramete. Per device might be set to 10 or 15 by default. We should be able to change it throught the device editor. And each parameter should have an update-rate parameter, set to 0 by default (that means to use the device value). If we set an update rate for a parameter, it will overwrite the device value.
From what I understand, setting an update-rate for an automation will became a very-low-priority issue.

how does it sound?

**Comment by bltzr** [comment]
good!

**Comment by jln-** [comment]
Sounds good to me as well !

**Comment by theod** [comment]
+1 for each of you. all those features sounds important to allow a precise data flow control from i-score.

this reminds me a discussion we had with @cbossut and @lossius in Albi : 

> For **Domain**, **BoundingMode**, **RepetitionFilter** and **RefreshRate** attributes, we can ask the following question “who is responsible ?” :
> - some devices are stupid (like a MIDI device).
> - some protocols have a bandwith.
> - some devices are slow to interpret the command (so we need to lookahead the value).
> - some devices are able to ramp them self and will tell you when they have ended it.

**Comment by bltzr** [comment]
good points !
As a start, I'd say that we should have a field for the RefreshRate in i-score (with a default value) - and then optionally, when devices are able to provide this info, we should use the value from the device (either globally on per-parameter)
outsourcing the ramp to the remote device is another, related, problem... but it seems much more complicated to handle...doesn't it ?

**Comment by bltzr** [comment]
@reno- I agree with you !

**Comment by theod** [comment]
> outsourcing the ramp to the remote device is another, related, problem... but it seems much more complicated to handle...doesn't it ?

sure ! but it was just to extrapolate about this issue as it is part of a global feature that needs to be designed in a wise way into the API.

**Comment by reno138** [comment]
Hey folks. Just a heads up...

I'm @reno and I'm pretty sure you're trying to tag @reno- 

Just wanted to make sure you are getting the right folks involved.

**Comment by jcelerier** [comment]
Mhh, everywhere on the conversation it looks like everybody wrote "@ reno-" ?  Maybe a github bug ?

**Comment by reno138** [comment]
Might be. Doesn't bother me. I just wanted to make sure you are getting the right people in on your conversation.

**Comment by reno-** [comment]
Not a github bug… I  changed the mistake when @reno sends the message. Sorry for the noise bro' !

**Comment by reno138** [comment]
No worries.

**Comment by jcelerier** [comment]
For now in dev I implemented a settings panel that allows to change the base execution rate.
- it's only set before starting the execution (so changing it mid-course won't work - for now).
- it's global and remembered between launches of i-score (it's not part of the project)

**Comment by theod** [comment]
great !
I've noticed that i-score reminds the last value even without no project loaded.
it is fine for the time being but to close this ticket I would say this value should be saved with the project : for example the boids_choregraphy project expect a 1 ms refresh rate in order to create 300 boids as quick as possible otherwise the creation loop misses many patterns.

**Comment by bltzr** [comment]
that's great ! Thanks @jcelerier !!!
if you can add the unit (i.e. ms) that would be more self-informative (I had to check if it was ms or fps)

**Comment by lossius** [comment]
Hi,

My two cents: I have also experienced network transmission overload and Max becoming unresponsive (spinning beach ball), so I agree that it will be useful to be able to set this as needed. At the same time I'd encourage a "convention over configuration" approach here:

1) i-score should have a sensible convention for default update rate. Personally I would suggest this to  be 5 ms. This could be set as a preference of the program, so that users could override it.

2) Each score could have a setable default update rate. By default this would be "default" and equal to the one in (1). If it is changed to something else, it overrides (1).

3) Each device within the score could have a device-specific update rate. Again, by default this would be "default" and equal to the one in (2). If it is changed to something else, it overrides (2).

4) Each event within the score could have a setable update rate. By default this would be "default" and equal to the one in (3). If it is changed to something else, it overrides (3).

**Comment by lossius** [comment]
And in following up: If the networked device is able of handling ramps itself, the responsibility of ramping could be delegated to the device rather than being done by i-score. This would be an additional setting relating to (2). Eventually, the question of whether the device is able to handle ramps or not could be part of the Minuit or OscQuery protocol itself, so that the user do not need to set it at all.

**Comment by jcelerier** [comment]
>  Eventually, the question of whether the device is able to handle ramps

How would it work in the case where i-score says "start ramp /x from 10 to 20 in five seconds", and then due to an interactive event, stops ramping at four seconds ? Does all "ramping" protocol support being interrupted ?

**Comment by bltzr** [comment]
yes.... and... what if the speed of the scenario or of the constraint holding the automation is increased ??... we would then probably need to access the clock of the remote ramping mechanism... so that's one more layer to address

**Comment by bltzr** [comment]
apart from that, we seem to all agree on what you propose, @lossius .
actually, 1 and 2 are already implemented, since the latest release :-) (thanks @jcelerier !)

**Comment by lossius** [comment]
> @jcelerier: How would it work in the case where i-score says "start ramp /x from 10 to 20 in five seconds", and then due to an interactive event, stops ramping at four seconds ? 

Assuming that a parameter is to ramp from 1.1 to 6.6 in 5 seconds, but is then stopped after 4 seconds, the external ramp can be terminated by a new message after 4 seconds that sets the value to what it is expected to be currently, namely 5.5

> @bltzr : yes.... and... what if the speed of the scenario or of the constraint holding the automation is increased ??... we would then probably need to access the clock of the remote ramping mechanism... so that's one more layer to address

This sounds messy to me and is probably not a good idea. If the speed has been increased already before the ramp is instantiated, the adjusted ramp time should be calculated by i-score before passing it to the receiving device. If the speed is being altered interactively _during_ the ramp, I guess the best solution is to temporarily override the external ramping mechanism, and instead drive the ramp from i-score.

**Comment by reno-** [comment]
> f the networked device is able of handling ramps itself, the responsibility of ramping could be delegated to the device rather than being done by i-score.

To be able to delegate the ramp, we first need to design a way to describe the automation. The default linear ramp is just one possibility as we can customise the automation in i-score

**Comment by cbossut** [comment]
For the rate of the automation, if I remember well, the API should propose
a way to use any scheduler for each process, and I think this option is
minimal complete and should be sufficient for the model.
Then, i-score GUI, could propose ways to change the rate by device or
parameter, or project.

2016-02-11 16:01 GMT+01:00 Trond Lossuis notifications@github.com:

> @jcelerier https://github.com/jcelerier: How would it work in the case
> where i-score says "start ramp /x from 10 to 20 in five seconds", and then
> due to an interactive event, stops ramping at four seconds ?
> 
> Assuming that a parameter is to ramp from 1.1 to 6.6 in 5 seconds, but is
> then stopped after 4 seconds, the external ramp can be terminated by a new
> message after 4 seconds that sets the value to what it is expected to be
> currently, namely 5.5
> 
> @bltzr https://github.com/bltzr : yes.... and... what if the speed of
> the scenario or of the constraint holding the automation is increased ??...
> we would then probably need to access the clock of the remote ramping
> mechanism... so that's one more layer to address
> 
> This sounds messy to me and is probably not a good idea. If the speed has
> been increased already before the ramp is instantiated, the adjusted ramp
> time should be calculated by i-score before passing it to the receiving
> device. If the speed is being altered interactively _during_ the ramp, I
> guess the best solution is to temporarily override the external ramping
> mechanism, and instead drive the ramp from i-score.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OSSIA/i-score/issues/147#issuecomment-182898735.

## 

Clément Bossut

**Comment by lossius** [comment]
> @reno- To be able to delegate the ramp, we first need to design a way to describe the automation. The default linear ramp is just one possibility as we can customise the automation in i-score

As i-score is using the JamomaCore C++ library, I would think that this should not be so much of a problem. If the automation curves that are available in i-score are adapted from the Jamoma FunctionLib, they will also be accessible to Jamoma-based devices, such as Jamoma for Max. In this case it is still enough to pass the necessary meta-information to recreate the curve in Max. That might, though, include specifying ramp function and some ramp function parameters before triggering the ramp itself.

On the other hand we also get the question whether such ramping curves can be standardised beyond Jamoma. In order to do so we would have to restrict ourselves to using something that is more or less standard in the field, such as Penners easing functions. I'm not necessarily convinced that this is desired, it could cut us of from using various curves that are very important for expressive means (such as breakpoint functions).

At the end of the day I believe that the question of delegating the ramp to the device rather than i-score is a question of _optimalisation_, in particular with respect to optimising network traffic. This might be a topic that can be addressed by other means as well, e.g. by bundling several updated values that are to be transmitted at the same time into one OSC bundle, rather then sending each of them as a separate OSC package.

**Comment by theod** [comment]
To bring a precision about what @cbossut said : in the API the gran guardian of the time is the TimeConstraint which inherits from Clock and have its own granularity (= update rate). A TimeConstraint can drive several TimeProcesses.

So if we want i-score to manage a per automation update rate, this will implies that, behind the scene, a "blue horizontal line" represents several TimeConstraints sharing the same start and end events, having the same durations and managing only one TimeProcess to manage a specific update rate for each of them.

This way the API stays really simple but I guess this will require a bit of work for @jcelerier and @nvuaille to reorganize what is managed by a "blue horizontal line" (?)

**Comment by theod** [comment]
about delegating a ramp : 
do we know other devices able to handle a ramp it self than Jamoma ?
I mean is there other programs that can be driven sending them an order to go to a value in an amount of time ?

**Comment by bltzr** [comment]
@lossius I'm not sure i-score is still using JamomaCore for managing ramps (AFAIK, this implied too much overhead for each segment, so it has been simplified to something more lightweight....)
@reno- anyway, there could probably be a way to translate i-score's power "coefficient" to Jamoma's... so that each segment of an i-score automation could be delegated to Jamoma 

@theod I'm not really convinced by having several constraints for different automation rates - why not using as the clock rate of the constraint the smallest common denominator of all rates of all automation it holds  
Concerning your last question, I don't know of any other environment that accepts ramps as OSC inputs, but I don't know many of them (especially not in details...)

**Comment by theod** [comment]
> @lossius  I'm not sure i-score is still using JamomaCore for managing ramps (AFAIK, this implied too much overhead for each segment, so it has been simplified to something more lightweight....)

yes Jamoma is mainly used in OSSIA API for the device management while ramps are based on C++ template architecture with 2 classes ([Curve](https://github.com/OSSIA/API/blob/master/Headers/Editor/Curve.h) and [CurveSegment](https://github.com/OSSIA/API/blob/master/Headers/Editor/CurveSegment.h))

notice I've just opened a discussion about the [relation between Jamoma and OSSIA](https://github.com/OSSIA/API/issues/27) to not pollute this thread.

**Comment by theod** [comment]
> @theod I'm not really convinced by having several constraints for different automation rates - why not using as the clock rate of the constraint the smallest common denominator of all rates of all automation it holds

first of all you need to understand this will be hidden for i-score user (it is just a matter of internal architecture).

Then I would say there are 2 approaches here : 
- having one clock at the faster rate means you need a filtering unit into each destination (I mean locally before to send values over the network). Those filtering can be based on value (like a repetition filter or a smooth filter) or based on time (like a speedlim, which means you deal with time in another place than into the scenario itself so it can result in some unexpected behavior). 
- having one clock per value we drive means we can avoid time filtering at the destination level before to send it over the network (and we could also imagine that value filtering could be a part of the automation).

so, to summarize the 2 options, we can produce more values than needed then filtering them before to send them or produce only values that need to be sent.

I prefer the second design but maybe some use cases can argue for the first one.

**Comment by theod** [comment]
> Concerning your last question, I don't know of any other environment that accepts ramps as OSC inputs, but I don't know many of them (especially not in details...)

I was asking this because if the feature is only related to control Jamoma parameters I think we need to imagine Jamoma as a library based on OSSIA API : in that way it means we are somehow talking about distributing scenario execution between several clients that can be driven by a master scenario server (which sounds related to something you already worked on @jcelerier, don't you ?)

**Comment by jcelerier** [comment]
> between several clients 

well, the premise was that these "several clients" were instances of i-score; if they are then there would be nothing particular to do and if they are not, I don't see any other way than case-by-case handling of such software's specificities if we want to advocate a specific support.

**Comment by bltzr** [comment]
I've skimmed through this issue (wow! that was a long one!), and just a thought:
now that we're beginning to have implementations of libossia for namespace managements, maybe delegating automations would be something to think about...
I guess that won't be happening before 2.0 but we can begin to specify it..

**Comment by bltzr** [comment]
It's even so long that I forgot that was about update rate, maybe that could be useful to create a new issue for automation delegation -> that's #462 !!!!

**Comment by bltzr** [comment]
update rates are now implemented in the parameters, on the libossia side

maybe what we need now, is to use this info to send the values from i-score at the closer (yet smaller) rate that the general update rate of score can deliver ?
would that make sense ?
